package common.protocol;

message Color4
{
  optional uint32 r = 1 [default = 0];
  optional uint32 g = 2 [default = 0];
  optional uint32 b = 3 [default = 0];
  optional uint32 a = 4 [default = 255];
}

message Vector2
{
  optional float x = 1;
  optional float y = 2;
}

message Vector3
{
  optional float x = 1;
  optional float y = 2;
  optional float z = 3;
}

message Vector4
{
  optional float x = 1;
  optional float y = 2;
  optional float z = 3;
  optional float w = 4;
}

message Matrix3x3
{
  optional Vector3 row_1 = 1;
  optional Vector3 row_2 = 2;
  optional Vector3 row_3 = 3;
}

message Matrix4x4
{
  optional Vector4 row_1 = 1;
  optional Vector4 row_2 = 2;
  optional Vector4 row_3 = 3;
  optional Vector4 row_4 = 4;
}

message FloatKeyframe
{
  optional int32 time = 1;
  optional float value = 2;

  // bezier control points
  optional int32 cp_in_time = 3;
  optional float cp_in_value = 4;

  optional int32 cp_out_time = 5;
  optional float cp_out_value = 6;
}

message FloatAnim
{
  optional uint32 type = 1;
  repeated FloatKeyframe keyframe = 2;
  optional float min_value = 3;
  optional float max_value = 4;
  optional float step = 5;
}

message Vector3Keyframe
{
//  option (common.protocol.context) = "void* context";
  optional int32 time = 1;
  optional Vector3 value = 2;
}

message Vector3Anim
{
  optional uint32 type = 1;
  repeated Vector3Keyframe keyframe = 2;
  optional Vector3 min_value = 3;
  optional Vector3 max_value = 4;
  optional Vector3 step = 5;
}
